README.md
[![build](https://github.com/LolzDEV/linux_system_info/actions/workflows/build.yml/badge.svg?branch=master)](https://github.com/LolzDEV/linux_system_info/actions/workflows/build.yml)
[![likes](https://badges.bar/linux_system_info/likes)](https://pub.dev/packages/linux_system_info/score)
[![popularity](https://badges.bar/linux_system_info/popularity)](https://pub.dev/packages/linux_system_info/score)
[![pub points](https://badges.bar/linux_system_info/pub%20points)](https://pub.dev/packages/linux_system_info/score)

## Linux System Info

A library for getting linux system information

## Usage

A simple usage example:

```dart
import 'package:linux_system_info/linux_system_info.dart';

void main() {

  //CPU

  var cpu_usage = CpuInfo.getCpuUsagePercentage(); // This returns cpu load in percentage; e.g. 2

  var cpu0_usage = CpuInfo.getProcessors()[0].getCpuUsagePercentage(); // This returns the cpu0 load in percentage (in case of multicore cpus); e.g. 2

  var cpu0_MHz = CpuInfo.getProcessors()[0].cpu_mhz; // You can also get a lot of information about the cpu cores; e.g. 3600

  //MEMORY

  var total_mem = MemInfo().mem_total_gb; // This returns the amount of RAM in GB (you can also get it in kb or mb); e.g. 16

  var total_swap = MemInfo().swap_total_gb; // This returns the amount of SWAP in GB (you can also get it in kb or mb) e.g. 2

  //SYSTEM

  var kernel_name = SystemInfo().kernel_name; // This returns the kernel name; e.g. 'Linux'

  var kernel_version = SystemInfo().kernel_version; // This returns the kernel version; e.g. '5.11.0-16-generic'

  var os_name = SystemInfo().os_name; // This returns the os name; e.g. 'Ubuntu'

  var os_version = SystemInfo().os_version; // This returns the os version; e.g. '21.04'

  var os_release = SystemInfo().os_release; // This value is a Map containing a parsed version of /etc/os-release

}

```

## Features and bugs

Please file feature requests and bugs at the [issue tracker][tracker].

[tracker]: https://github.com/LolzDEV/linux_system_info/issues


CHANGELOG.md
## 0.0.7

- Added GnomeInfo and GPUInfo

## 0.0.6

- Minor fixes

## 0.0.5

- Moved to null safety

## 0.0.4

- Formatted code with `dart format`

## 0.0.3

- Added documentation comments

## 0.0.2

- Initial version, created by LolzDEV


test/linux_system_info_test.dart
import 'package:linux_system_info/linux_system_info.dart';
import 'package:test/expect.dart';
import 'package:test/scaffolding.dart';

void main() {
  test('Check returned data', () {
    var cpu0_MHz = CpuInfo.getProcessors()[0].cpu_mhz;
    expect(cpu0_MHz.runtimeType, equals(double));

    var total_mem = MemInfo().mem_total_gb;
    expect(total_mem.runtimeType, equals(int));

    var total_swap = MemInfo().swap_total_gb;
    expect(total_swap.runtimeType, equals(int));

    var kernel_name = SystemInfo().kernel_name;
    expect(kernel_name.runtimeType, equals(String));

    var kernel_version = SystemInfo().kernel_version;
    expect(kernel_version.runtimeType, equals(String));

    var os_name = SystemInfo().os_name;
    expect(os_name.runtimeType, equals(String));

    var os_version = SystemInfo().os_version;
    expect(os_version.runtimeType, equals(String));

    var os_release = SystemInfo().os_release;
    expect(os_release.runtimeType, equals(<String, String>{}.runtimeType));
  });
}


lib/src/linux_system_info_base.dart
import 'dart:ffi';
import 'dart:io';
import 'package:dbus/dbus.dart';
import 'package:ffi/ffi.dart';
import 'package:path/path.dart' as p;
import 'package:tuple/tuple.dart';
import 'package:xdg_directories/xdg_directories.dart' as xdg;
import 'package:xml/xml.dart';

/// Class for handling general system information
class SystemInfo {
  /// Current kernel name, e.g.: `Linux`
  String kernel_name = '';

  /// Current kernel version, e.g.: `5.11.0-16-generic`
  String kernel_version = '';

  /// Current os name, e.g.: `Ubuntu`
  String os_name = '';

  /// Current os version, e.g.: `21.04`
  String os_version = '';

  /// A parsing of `/etc/os-release` containing information like the distro codename and homepage
  Map<String, String> os_release = <String, String>{};

  /// Empty constructor, this loads all the information from `/proc/version` and from `/etc/os-release`
  SystemInfo() {
    var content = File('/proc/version').readAsStringSync();
    kernel_name = content.split(' ')[0];
    kernel_version = content.split(' ')[2];
    var lines = File('/etc/os-release').readAsLinesSync();
    os_release = <String, String>{};
    for (var e in lines) {
      var data = e.split('=');
      os_release[data[0]] = data[1].replaceAll('\"', '');
    }

    os_name = os_release['NAME'] ?? '';
    os_version = os_release['VERSION_ID'] ?? '';
  }
}

/// Class containing CPU core data
class Cpu {
  /// Core number
  int processor = 0;

  /// CPU vendor identifier
  String vendor_id = '';

  /// CPU family
  int cpu_family = -1;

  /// CPU model
  int model = -1;

  /// CPU model name
  String model_name = '';

  /// CPU stepping
  int stepping = -1;

  /// CPU microcode
  int microcode = -1;

  /// CPU MHz
  double cpu_mhz = -1.0;

  /// CPU cache size
  String cache_size = '';

  /// CPU physical id
  int physical_id = -1;

  /// CPU siblings
  int siblings = -1;

  /// CPU core identifier
  int core_id = -1;

  /// Total amount of cores
  int cpu_cores = -1;

  /// CPU apicid
  int apicid = -1;

  /// CPU initial apicid
  int initial_apicid = -1;

  /// If cpu has an FPU (Floating point unit)
  bool fpu = false;

  /// If FPU has an exception
  bool fpu_exception = false;

  /// Cpuid level
  int cpuid_level = -1;
  bool wp = false;

  /// CPU's flags
  List<String> flags = [];

  /// CPU bogomips
  double bogomips = -1.0;

  /// CPU clflush's size
  int clflush_size = -1;

  /// CPU cache alignment
  int cache_alignment = -1;
  Tuple2<String, String> address_sizes = Tuple2<String, String>('', '');

  /// Constructor for creating a [Cpu] with data in it
  Cpu.fromData(
      this.processor,
      this.vendor_id,
      this.cpu_family,
      this.model,
      this.model_name,
      this.stepping,
      this.microcode,
      this.cpu_mhz,
      this.cache_size,
      this.physical_id,
      this.siblings,
      this.core_id,
      this.cpu_cores,
      this.apicid,
      this.initial_apicid,
      this.fpu,
      this.fpu_exception,
      this.cpuid_level,
      this.wp,
      this.flags,
      this.bogomips,
      this.clflush_size,
      this.cache_alignment,
      this.address_sizes);

  /// Empty constructor, creates a [Cpu] object with default values in it
  Cpu();

  /// Get current core load
  List<int> load() {
    var f = File('/proc/stat');
    var lines = f.readAsLinesSync();

    var loads = lines[processor + 1].split(' ');

    loads.removeAt(0);

    var total = loads
        .map((String token) => int.parse(token))
        .toList()
        .reduce((int a, int b) => a + b);

    var idle = int.parse(loads[3]);

    return [idle, total];
  }

  /// Get a [Stream] for handling the core usage in percentage as a [double]
  Stream<double> getCpuUsagePercentage() async* {
    var idleTotalPrev = <int>[0, 0];

    while (true) {
      var idleTotal = load();
      var dTotal = idleTotal[0] - idleTotalPrev[0];
      var dLoad = idleTotal[1] - idleTotalPrev[1];
      idleTotalPrev = idleTotal;

      var percent = 100.0 * (1.0 - dTotal / dLoad);
      yield percent;

      sleep(Duration(seconds: 1));
    }
  }
}

/// Class memory information
class MemInfo {
  /// Total RAM amount in Kb
  int mem_total = 0;

  /// Total RAM amount in Mb
  int mem_total_mb = 0;

  /// Total RAM amount in Gb
  int mem_total_gb = 0;

  /// Free RAM amount in Kb
  int mem_free = 0;

  /// Free RAM amount in Mb
  int mem_free_mb = 0;

  /// Free RAM amount in Gb
  int mem_free_gb = 0;

  /// Total Swap amount in Kb
  int swap_total = 0;

  /// Total Swap amount in Mb
  int swap_total_mb = 0;

  /// Total Swap amount in Gb
  int swap_total_gb = 0;

  /// Free Swap amount in Kb
  int swap_free = 0;

  /// Free Swap amount in Mb
  int swap_free_mb = 0;

  /// Free Swap amount in Gb
  int swap_free_gb = 0;

  /// Empty constructor, this loads all the information from `/proc/meminfo`
  MemInfo() {
    var lines = File('/proc/meminfo').readAsLinesSync();

    for (var e in lines) {
      var data = e.split(':');
      if (data.length == 2) {
        data[0] = data[0].trim();
        switch (data[0]) {
          case 'MemTotal':
            {
              data[1] = data[1].trim().substring(0, data[1].trim().length - 3);
              mem_total = int.parse(data[1]);
              mem_total_mb = (mem_total / 1024).round();
              mem_total_gb = (mem_total_mb / 1024).round();
            }
            break;
          case 'MemFree':
            {
              data[1] = data[1].trim().substring(0, data[1].trim().length - 3);
              mem_free = int.parse(data[1]);
              mem_free_mb = (mem_free / 1024).round();
              mem_free_gb = (mem_free_mb / 1024).round();
            }
            break;
          case 'SwapTotal':
            {
              data[1] = data[1].trim().substring(0, data[1].trim().length - 3);
              swap_total = int.parse(data[1]);
              swap_total_mb = (swap_total / 1024).round();
              swap_total_gb = (swap_total_mb / 1024).round();
            }
            break;
          case 'SwapFree':
            {
              data[1] = data[1].trim().substring(0, data[1].trim().length - 3);
              swap_free = int.parse(data[1]);
              swap_free_mb = (swap_free / 1024).round();
              swap_free_gb = (swap_free_mb / 1024).round();
            }
            break;
        }
      }
    }
  }
}

/// Class containing CPU information
class CpuInfo {
  /// Get a [List] containing all cpu's cores in a [Cpu] class
  static List<Cpu> getProcessors() {
    var cpus = <Cpu>[];
    var curCpu = 0;
    var currentCpu = Cpu();
    var lines = File('/proc/cpuinfo').readAsLinesSync();
    for (var e in lines) {
      var data = e.split(':');

      if (data.isNotEmpty) {
        data[0] = data[0].trim();
        if (data[0] == 'processor' && curCpu > 0) {
          cpus.add(currentCpu);
          currentCpu = Cpu();
          currentCpu.processor = int.parse(data[1]);
          curCpu += 1;
        }

        switch (data[0]) {
          case 'vendor_id':
            {
              currentCpu.vendor_id = data[1];
            }
            break;
          case 'cpu family':
            {
              currentCpu.cpu_family = int.parse(data[1]);
            }
            break;
          case 'model':
            {
              currentCpu.model = int.parse(data[1]);
            }
            break;
          case 'model name':
            {
              currentCpu.model_name = data[1];
            }
            break;
          case 'stepping':
            {
              currentCpu.stepping = int.parse(data[1]);
            }
            break;
          case 'microcode':
            {
              currentCpu.microcode = int.parse(data[1]);
            }
            break;
          case 'cpu MHz':
            {
              currentCpu.cpu_mhz = double.parse(data[1]);
            }
            break;
          case 'cache size':
            {
              currentCpu.cache_size = data[1];
            }
            break;
          case 'physical id':
            {
              currentCpu.physical_id = int.parse(data[1]);
            }
            break;
          case 'siblings':
            {
              currentCpu.siblings = int.parse(data[1]);
            }
            break;
          case 'core id':
            {
              currentCpu.core_id = int.parse(data[1]);
            }
            break;
          case 'cpu cores':
            {
              currentCpu.cpu_cores = int.parse(data[1]);
            }
            break;
          case 'apicid':
            {
              currentCpu.apicid = int.parse(data[1]);
            }
            break;
          case 'initial apicid':
            {
              currentCpu.initial_apicid = int.parse(data[1]);
            }
            break;
          case 'fpu':
            {
              currentCpu.fpu = data[1] == 'yes' ? true : false;
            }
            break;
          case 'fpu_exception':
            {
              currentCpu.fpu_exception = data[1] == 'yes' ? true : false;
            }
            break;
          case 'cpuid level':
            {
              currentCpu.cpuid_level = int.parse(data[1]);
            }
            break;
          case 'wp':
            {
              currentCpu.wp = data[1] == 'yes' ? true : false;
            }
            break;
          case 'flags':
            {
              currentCpu.flags = data[1].split(' ');
            }
            break;
          case 'bogomips':
            {
              currentCpu.bogomips = double.parse(data[1]);
            }
            break;
          case 'clflush size':
            {
              currentCpu.clflush_size = int.parse(data[1]);
            }
            break;
          case 'cache_alignment':
            {
              currentCpu.cache_alignment = int.parse(data[1]);
            }
            break;
          case 'address sizes':
            {
              currentCpu.address_sizes =
                  Tuple2(data[1].split(',')[0], data[1].split(',')[1]);
            }
            break;
        }

        if (curCpu == 0) {
          curCpu += 1;
        }
      }
    }
    return cpus;
  }

  /// Get total cpu load
  static List<int> load() {
    var f = File('/proc/stat');
    var lines = f.readAsLinesSync();

    var loads = lines[0]
        .substring('cpu  '.length)
        .split(' ')
        .map((String token) => int.parse(token))
        .toList();

    var idle = loads[3];
    var total = loads.reduce((int a, int b) => a + b);

    return [idle, total];
  }

  /// Get a [Stream] for handling the total cpu usage in percentage as a [double]
  static Stream<double> getCpuUsagePercentage() async* {
    var idleTotalPrev = <int>[0, 0];

    while (true) {
      var idleTotal = load();
      var dTotal = idleTotal[0] - idleTotalPrev[0];
      var dLoad = idleTotal[1] - idleTotalPrev[1];
      idleTotalPrev = idleTotal;

      var percent = 100.0 * (1.0 - dTotal / dLoad);
      yield percent;

      sleep(Duration(seconds: 1));
    }
  }
}

class Gpu {
  /// The model name of the GPU e.g. "Intel(R) HD Graphics"
  String model = '';

  Gpu(this.model);
}

/// GPU information
class GpuInfo {
  /// Fetches GPU information from:
  /// - `net.hadess.SwitcherooControl.GPUs` (D-Bus)
  /// - `org.gnome.SessionManager.Renderer` (D-Bus)
  /// - `GL_RENDERER` (OpenGL)
  static Future<List<Gpu>> load() async {
    final gpus = await _getSwitcherooGpus() ??
        await _getGnomeRenderer() ??
        await _getGlRenderer();
    return gpus?.map((gpu) => Gpu(gpu)).toList() ?? [];
  }

  static Future<DBusValue?> _getDBusProperty({
    required DBusClient client,
    required String interface,
    required String path,
    required String property,
  }) async {
    try {
      final object = DBusRemoteObject(
        client,
        name: interface,
        path: DBusObjectPath(path),
      );
      final value = await object.getProperty(interface, property);
      await client.close();
      return value;
    } on DBusErrorException catch (_) {
      return null;
    }
  }

  static Future<Iterable<String>?> _getSwitcherooGpus() async {
    final value = await _getDBusProperty(
      client: DBusClient.system(),
      interface: 'net.hadess.SwitcherooControl',
      path: '/net/hadess/SwitcherooControl',
      property: 'GPUs',
    );
    if (value == null) return null;

    final gpus =
        (value as DBusArray).children.map((child) => child.toNative()).toList();
    gpus.sort((a, b) {
      if (a['Default'] == true) return 1;
      if (b['Default'] == true) return -1;
      return 0;
    });
    return gpus.map((gpu) => gpu['Name'] as String);
  }

  static Future<Iterable<String>?> _getGnomeRenderer() async {
    final value = await _getDBusProperty(
      client: DBusClient.system(),
      interface: 'org.gnome.SessionManager',
      path: '/org/gnome/SessionManager',
      property: 'Renderer',
    );
    if (value == null) return null;
    return <String>[(value as DBusString).value];
  }

  static Future<Iterable<String>?> _getGlRenderer() async {
    try {
      final lib = DynamicLibrary.open('libGL.so.1');
      final glGetString =
          lib.lookupFunction<_GlGetStringC, _GlGetStringDart>('glGetString');
      const GL_RENDERER = 0x1F01;
      final cstr = glGetString(GL_RENDERER);
      if (cstr == nullptr) return null;
      return <String>[cstr.cast<Utf8>().toDartString()];
    } on ArgumentError catch (_) {
      return null;
    }
  }
}

typedef _GlGetStringC = Pointer<Uint8> Function(Uint32 name);
typedef _GlGetStringDart = Pointer<Uint8> Function(int name);

class GnomeInfo {
  GnomeInfo._([this.version = '', this.distributor = '']);

  final String version;
  final String distributor;

  static GnomeInfo? _instance;

  factory GnomeInfo() {
    _instance ??= _init();
    return _instance!;
  }

  static GnomeInfo _init() {
    final file = _findDataFile('gnome/gnome-version.xml');
    if (file == null) return GnomeInfo._();
    final xml = file.readAsStringSync();
    return _parseGnomeVersion(xml);
  }

  static GnomeInfo _parseGnomeVersion(String xml) {
    final doc = XmlDocument.parse(xml);

    final elements = <String, String>{
      for (final element in doc.rootElement.childElements)
        element.name.toString(): element.text,
    };

    final version = <String?>[
      elements['platform'],
      elements['minor'],
      elements['micro'],
    ].takeWhile((v) => v != null).join('.');

    final distro = elements['distributor'];

    return GnomeInfo._(version, distro ?? '');
  }

  static File? _findDataFile(String path) {
    for (final dir in xdg.dataDirs) {
      final file = File(p.join(dir.path, path));
      if (file.existsSync()) {
        return file;
      }
    }
    return null;
  }
}


example/linux_system_info_example.dart
import 'package:linux_system_info/linux_system_info.dart';

Future<void> main() async {
  //CPU

  var cpu_usage = CpuInfo
      .getCpuUsagePercentage(); // This returns cpu load in percentage; e.g. 2

  var cpu0_usage = CpuInfo.getProcessors()[0]
      .getCpuUsagePercentage(); // This returns the cpu0 load in percentage (in case of multicore cpus); e.g. 2

  var cpu0_MHz = CpuInfo.getProcessors()[0]
      .cpu_mhz; // You can also get a lot of information about the cpu cores; e.g. 3600

  //MEMORY

  var total_mem = MemInfo()
      .mem_total_gb; // This returns the amount of RAM in GB (you can also get it in kb or mb); e.g. 16

  var total_swap = MemInfo()
      .swap_total_gb; // This returns the amount of SWAP in GB (you can also get it in kb or mb) e.g. 2

  //GPU

  var gpu_model = (await GpuInfo.load())[0]
      .model; // This returns the model name of the GPU e.g. "Intel HD Graphics"

  //SYSTEM

  var kernel_name =
      SystemInfo().kernel_name; // This returns the kernel name; e.g. 'Linux'

  var kernel_version = SystemInfo()
      .kernel_version; // This returns the kernel version; e.g. '5.11.0-16-generic'

  var os_name = SystemInfo().os_name; // This returns the os name; e.g. 'Ubuntu'

  var os_version =
      SystemInfo().os_version; // This returns the os version; e.g. '21.04'

  var os_release = SystemInfo()
      .os_release; // This value is a Map containing a parsed version of /etc/os-release

  //GNOME
  var gnome_version =
      GnomeInfo().version; // This returns the Gnome version; e.g. '3.38.5'

  var gnome_distributor = GnomeInfo()
      .distributor; // This returns the Gnome distributor; e.g. 'Ubuntu'
}
